### ISSUE TITLE
UnwindSafe docs are unclear

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang/rust#65717]
After reading the current docs of `panic::UnwindSafe` I am left with the feeling that the explanation of the concept of "panic safety" and its auto trait has only been handwaved. The docs point to the RFC, which does a better job, but I still do not feel like I fully understood what "panic safety" is, what does `UnwindSafe` convey, and when is it ok to use `AssertUnwindSafe` or not. 

I feel like the docs could do a much better job at explaining all of this, and that some examples in the RFC would help, but I don't think that any documentation we currently have about this is enough.

This is a complex topic and I have no experience teaching it to others, but if somebody has done that, their feedback should probably be taken into account. 

I think the API docs are very important to get right because they are currently the only teaching materials that we have about this aspect of the language. The Rust book does not cover this either, and the nomicon [Exception safety](https://doc.rust-lang.org/nomicon/exception-safety.html?highlight=panic,safety#exception-safety) has very few words to say about this.

cc @steveklabnik @rust-lang/docs

**Comment by steveklabnik** [comment]
I haven't actually looked at these docs, so I can't agree or disagree :) Thanks for opening the issue. I wish that there was something more... actionable, about it, but this is a good first step 👍

**Comment by gnzlbg** [comment]
> Thanks for opening the issue. I wish that there was something more... actionable, about it, but this is a good first step

Yes, me too. I'm not sure how to word what I got from reading these docs. I think my first intention is whether "it is just me" or whether other people have trouble understanding these.

The docs aren't short, there are a lot of words in there, and a lot of things are explained. By "unclear" I I think I more precisely mean that, after reading the docs, I cannot answer basic questions about what I think the docs attempted to explain:

* What is panic safety? (I'm missing a concise definition of panic safety - I find the definitions here and in the RFC "vague")
* Why is panic safety important? (IIUC, a program that's not "panic safe" can still be "safe" as long as it does not contain `unsafe` code, so... "panic un-safety" is not "unsafe" per se, but it could cause some `unsafe` code that makes some assumptions about panic safety to be unsound) 
* What problem does "panic safety", `AssertUnwindSafe`, `UnwindSafe`, etc. solve ?
* When should I implement / not implement the `UnwindSafe` trait for a type ? 
* Which types in the standard library are not `UnwindSafe` and why? (I think examples of already existing types would have helped me understand this better, e.g., by not only reading the definitions but also seeing them being used in the wild - the RFC contains some more examples and these helped)
* When is it ok to use `AssertUnwindSafe` ? (IIUC if some type is not panic safe, but that does not pose a problem in the context it is used, then one can use this escape hatch, but see #40628 )

There is also the issue that one needs to click through a lot of pieces to obtain the whole picture (e.g. the UnwindSafe trait docs, the AssertUnwindSafe docs, the RFC, the nomicon, etc.). I think it might be better to cover panic safety holistically in the `core::panic` top-level module docs, and explain there what it is, and which problems do the different pieces solve.

Another reason why the documentation feels unclear is because there is actually a lot of content in there, but reading it doesn't feel "definitive". E.g. it feels as if it isn't clear at all what "panic safety" is, nor what problems do these utilities solve, if any, and / or whether they are good solutions to these problems. I'm not sure how else to explain it, but I run into this docs once a year, and I've never felt I truly understood what the point of these features are.  #40628 mentions 

>  I've seen now quite a bit of people that told me that "they just wrap everything in AssertUnwindSafe" defeating the whole purpose of it.

and... this is what I ended up doing last time, and what I ended up doing today. After reading the docs, I'm not even able to write a comment saying whether doing that is "ok". In January or so I remember running into this again here: https://github.com/rust-lang/rust/blob/d6e4028a0db6b13d9a603baad109d6c902802c03/src/libtest/lib.rs#L570 and I was like, "is this code correct?". Hard to tell, since there is no comment about why `AssertUnwindSafe` is correct there. I tried to write a comment, but if you take a look the function that gets passed to `catch_unwind` is a generic closure, so... it can be anything. How can one know that it is definitely going to be unwind safe ?

**Comment by jakoschiko** [comment]
@gnzlbg @BurntSushi @AltSysrq

If

- libcore is doing `catch_unwind(AssertUnwindSafe(testfn))` and
- quickcheck is doing [`panic::catch_unwind(panic::AssertUnwindSafe(fun))`](https://github.com/BurntSushi/quickcheck/blob/7845a988894e3e02233bf2115b3d5ce91a914ad9/src/tester.rs#L415) and
- proptest is doing [`panic::catch_unwind(AssertUnwindSafe(|| test(case)))`](https://github.com/AltSysrq/proptest/blob/dfe3353a9e9d1ce33f12aeb8ecf4f082b480d4c2/proptest/src/test_runner/runner.rs#L246),

then it must be correct, right?

**Comment by gnzlbg** [comment]
@jakoschiko I think all of those libraries use `AssertUnwindSafe` incorrectly. Consider:

```rust
lazy_static! { static ref V: Mutex<Vec<u8>> = Mutex::new(vec![]); }

#[should_panic] #[test] fn foo() { unsafe {
        { let mut v = V.lock().unwrap(); v.reserve(42); v.set_len(42); }
        panic!();
    }
}
#[test] fn bar() {
        let v = V.lock().unwrap(); let v: &Vec<u8> = &*v;
        for i in v {    dbg!(*i); }
}
```

Here if `foo` is run before `bar`, `V` is left in an invalid state, and when `bar` runs, it will use a `Vec` with broken invariants, resulting in undefined behavior. 

The whole point of `UnwindSafe` is preventing these programs from compiling. Apparently, "correct usage"  of `AssertUnwindSafe` requires its user to prove that the closure is actually `UnwindSafe`. Those test libraries take arbitrary user tests and their usage of `AssertUnwindSafe` is incorrect because they do not prove that these tests do not do anything incorrect. What all those test libraries should do instead is require a `+ UnwindSafe` bound on the user code they execute. Ideally, that would prevent the example above from compiling, e.g., because `foo` captures a `V: &mut Mutex<...Vec<...>>` type that is not `UnwindSafe`.

**Comment by sfackler** [comment]
UnwindSafe is not a "real" safety barrier- `AssertUnwindSafe` does not require `unsafe` to use! The `foo` test in your example is the problem, not the use of catch_unwind. Even if the test harness did not use catch_unwind at all, you'd still run into exactly the same problem as long as the harness was running more than one thread.

**Comment by gnzlbg** [comment]
@sfackler What you mention is covered in the `UnwindSafe` docs, which say that the goal of `UnwindSafe` is to prevent accidental misuses like this. The question here is whether `libtest` and other libraries are using `AssertUnwindSafe` correctly by stamping it on `catch_unwind` calls to **unknown code**. 

These libraries don't know what that code does, so why are they asserting that such code is `UnwindSafe` ?

**Comment by sfackler** [comment]
`AssertUnwindSafe` is also there to say "I don't care about unwind safety, just make this thing work".

**Comment by gnzlbg** [comment]
@sfackler if that's the case, then the uses in libstd (and by the people in #40628 that just stamp `AssertUnwindSafe` everywhere) are correct, and the `AssertUnwindUnsafe` docs are not only unclear, but also completely incorrect. Currently, the docs that `AssertUnwindSafe` is there to _assert_ something, e.g., in:

> Wrapping the entire closure amounts to a blanket assertion that all captured variables are unwind safe.

which is a very different thing from expressing that one does not care about unwind safety.

**Comment by hkBst** [comment]
@gnzlbg If you are asserting that some unknown code is unwind-safe, then that is tantamount to saying that you don't care about the unwind-safety of the unknown code, or, more charitably, that there is no unwind-safety to worry about for that particular use of catch_unwind.

The relevant docs were moved 4 years ago, but seem mostly unchanged since, and I suspect before then. I've briefly read over them and the RFC, but failed to spot anything in need of improvement. Perhaps you can take another look at them and see if you can clarify how they are unclear or can be improved?
